scripts/globals/homepoint.lua

```diff
@@ -1,361 +1,248 @@
+		local HpTeleportMask3 = bit.bor( bit.lshift( player:getVar("HpTeleportMask3a"), 16), player:getVar("HpTeleportMask3b"));
+		local HpTeleportMask4 = bit.bor( bit.lshift( player:getVar("HpTeleportMask4a"), 16), player:getVar("HpTeleportMask4b"));
+
+		-- Register new homepoint?
+		local newHp = 0;
+		if( homepoints[hpIndex][1] == 1) then
+			if( bit.rshift( bit.lshift( HpTeleportMask1, 32 - homepoints[hpIndex][2]), 31) == 0) then
+				newHp = 65536; -- This value causes the "You have registered a new home point!" dialog to display
+				
+				-- Update the homepoint teleport mask with the new location
+				HpTeleportMask1 = bit.bor( HpTeleportMask1, bit.lshift( 1, homepoints[hpIndex][2] - 1)); 
+
+				-- Save new mask to database
+				player:setVar("HpTeleportMask1a", bit.rshift( HpTeleportMask1, 16));
+				player:setVar("HpTeleportMask1b", bit.rshift( bit.lshift( HpTeleportMask1, 16), 16));
+			end
+		elseif(  homepoints[hpIndex][1] == 2) then
+			if( bit.rshift( bit.lshift( HpTeleportMask2, 32 - homepoints[hpIndex][2]), 31) == 0) then
+				newHp = 65536;
+				
+				HpTeleportMask2 = bit.bor( HpTeleportMask2, bit.lshift( 1, homepoints[hpIndex][2] - 1)); 
+
+				player:setVar("HpTeleportMask2a", bit.rshift( HpTeleportMask2, 16));
+				player:setVar("HpTeleportMask2b", bit.rshift( bit.lshift( HpTeleportMask2, 16), 16));
+			end
+		elseif(  homepoints[hpIndex][1] == 3) then
+			if( bit.rshift( bit.lshift( HpTeleportMask3, 32 - homepoints[hpIndex][2]), 31) == 0) then
+				newHp = 65536;
+				
+				HpTeleportMask3 = bit.bor( HpTeleportMask3, bit.lshift( 1, homepoints[hpIndex][2] - 1)); 
+
+				player:setVar("HpTeleportMask3a", bit.rshift( HpTeleportMask3, 16));
+				player:setVar("HpTeleportMask3b", bit.rshift( bit.lshift( HpTeleportMask3, 16), 16));
+			end
+		elseif(  homepoints[hpIndex][1] == 4) then
+			if( bit.rshift( bit.lshift( HpTeleportMask4, 32 - homepoints[hpIndex][2]), 31) == 0) then
+				newHp = 65536;
+				
+				HpTeleportMask4 = bit.bor( HpTeleportMask4, bit.lshift( 1, homepoints[hpIndex][2] - 1)); 
+
+				player:setVar("HpTeleportMask4a", bit.rshift( HpTeleportMask4, 16));
+				player:setVar("HpTeleportMask4b", bit.rshift( bit.lshift( HpTeleportMask4, 16), 16));
+			end
+		end
+
+		player:startEvent( csid, 0, HpTeleportMask1, HpTeleportMask2, HpTeleportMask3, HpTeleportMask4, player:getGil(), 4095, hpIndex + newHp);
+	else
+		player:PrintToPlayer( "Home point teleports are currently disabled on this server.");
+		player:startEvent( csid, 0, 0, 0, 0, 0, player:getGil(), 4095, hpIndex);
 	end
-	if(mask == 3) then
-		local hps = player:getVar("hpmask3")
-		if (bit.band(hps, bit.lshift(1, (num))) == 0)then
-			player:setVar("hpmask3",bit.bor(hps, bit.lshift(1, (num))))
-			return 65536
-		end
+end;
+
+function hpTeleport( player, option)
+
+	if( option == 2 or option > 65536 and option < 8323075) then
+		local hpIndex = (option - 2) / 65536; -- Calculate hpIndex based on option selected
```

Comment:
use bit operations instead of dividing (rshift I think is what you're looking for here)